### ISSUE TITLE
SFTP connection not working with OpenSSH server/Debian 11

### ISSUE THREAD

**Issue by thwerks** [PhilippC/keepass2android#2298]
Im trying to open a database file over an ssh connection but after i enter the information its hangs for a few seconds then drops back to the main screen where i ether create or open a database. I dont see any errors, it just closes the window and goes back to the initial screen and doesnt open the browser to select my file.

Im connecting to an OpenSSH server, its debian 11 with default encryption settings.  I've obviously added my private key to the ssh servers authorized keys, password authentication also fails.  The other android apps i have like ES File Explorer have no problem connecting to my multiple devices using OpenSSH with a private/public key.

I dont see anything wrong with my configuration, everything looks correct, other apps work fine. How do i figure out whats going wrong? 

Keepass ver 1.09e-r1

**Comment by PhilippC** [comment]
can you enable debug logging in the app's settings and see if that contains some hints (after trying again)?

**Comment by waltersdorff** [comment]
Just piggybacking sinc I have the same issue with connection via SFTP to my NAS Storage. SFTP works flawless with other apps through Android (Cx Explorer) or WinSCP on Windows. 

Here is some of the debug log, I changed the user, password, domain name

08.04.2023 08:40:54:392 -- FileSelect.OnCreate
08.04.2023 08:40:54:405 -- FileSelect.OnStart
08.04.2023 08:40:54:406 -- FileSelect.OnResume
08.04.2023 08:40:54:412 -- SelectCurrentDbActivity.OnStop 7
08.04.2023 08:40:54:883 -- PasswordActivity.OnStop 8
08.04.2023 08:40:54:885 -- PasswordActivity.OnDestroyTrue 8
08.04.2023 08:40:58:718 -- FileSelect.OnPause
08.04.2023 08:40:58:727 -- SelectStorageLocationActivity.OnCreate
08.04.2023 08:40:59:270 -- FileSelect.OnStop
08.04.2023 08:41:01:301 -- onAR
08.04.2023 08:41:01:302 -- base.onAR
08.04.2023 08:41:01:307 -- StartSelectFile sftp
08.04.2023 08:41:31:249 -- PasswordActivity.OnCreate 9
08.04.2023 08:41:31:250 -- PasswordActivity:apptask= 9
08.04.2023 08:41:31:265 -- GetIocFromLaunchIntent()
08.04.2023 08:41:31:265 -- no keyprovider specified
08.04.2023 08:41:31:266 -- Reset keyfile
08.04.2023 08:41:31:268 -- PasswordActivity.OnStart 9
08.04.2023 08:41:31:268 -- PasswordActivity.OnResume 9
08.04.2023 08:41:31:269 --  DB null 9
08.04.2023 08:41:31:269 -- starting: True, Finishing: False, _performingLoad: False
08.04.2023 08:41:31:269 -- sftp://user:password@sample_domain.eu:2222/Folder/Database.kdbx isCached = False
08.04.2023 08:41:31:270 -- Pre-loading database file starting
08.04.2023 08:41:31:271 -- sftp://user:password@sample_domain.eu:2222/Folder/Database.kdbx isCached = False
08.04.2023 08:41:31:271 -- CFS: OpenWhenNoLocalChanges
08.04.2023 08:41:31:303 -- PasswordModeSpinner item selected: 0
08.04.2023 08:41:31:332 -- FileSelect.OnDestroyTrue
08.04.2023 08:41:31:615 -- Creating application keepass2android.keepass2android. Version=189
08.04.2023 08:41:31:648 -- SelectCurrentDbActivity.OnCreate 0
08.04.2023 08:41:31:649 -- SelectCurrentDbActivity:apptask= 0
08.04.2023 08:41:31:664 -- Loaded task keepass2android.NullTask
08.04.2023 08:41:31:664 -- Task in activity SelectCurrentDbActivity 0 changed to NullTask
08.04.2023 08:41:31:672 -- SelectCurrentDbActivity.OnStart 0
08.04.2023 08:41:31:677 -- SelectCurrentDbActivity 0: OnActivityResult Canceled/1
08.04.2023 08:41:31:677 -- TryGetFromActivityResult: no data
08.04.2023 08:41:31:678 -- Task in activity SelectCurrentDbActivity 0 changed to NullTask
08.04.2023 08:41:31:679 -- Finishing SelectCurrentDbActivity 0
08.04.2023 08:41:31:682 -- SelectCurrentDbActivity.OnResume 0
08.04.2023 08:41:31:682 --  DB null 0
08.04.2023 08:41:31:682 -- SelectCurrentDbActivity.OnResume 0
08.04.2023 08:41:31:683 --  DB null 0
08.04.2023 08:41:31:684 -- Finishing SelectCurrentDbActivity 0
08.04.2023 08:41:31:685 -- SelectCurrentDbActivity.OnPause 0
08.04.2023 08:41:32:90 -- SelectCurrentDbActivity.OnStop 0
08.04.2023 08:41:32:98 -- SelectCurrentDbActivity.OnDestroyTrue 0
08.04.2023 08:41:34:730 -- KeePass.OnCreate 1
08.04.2023 08:41:34:730 -- KeePass:apptask= 1
08.04.2023 08:41:34:731 -- Loaded task keepass2android.NullTask
08.04.2023 08:41:34:731 -- Task in activity KeePass 1 changed to NullTask
08.04.2023 08:41:34:731 -- KeePass.OnCreate
08.04.2023 08:41:34:733 -- KeePass.OnStart 1
08.04.2023 08:41:34:733 -- KeePass.OnStart
08.04.2023 08:41:34:746 -- SelectCurrentDbActivity.OnCreate 2
08.04.2023 08:41:34:746 -- SelectCurrentDbActivity:apptask= 2
08.04.2023 08:41:34:750 -- Loaded task keepass2android.NullTask
08.04.2023 08:41:34:750 -- Task in activity SelectCurrentDbActivity 2 changed to NullTask
08.04.2023 08:41:34:750 -- SelectCurrentDbActivity.OnStart 2
08.04.2023 08:41:34:752 -- SelectCurrentDbActivity.OnResume 2
08.04.2023 08:41:34:752 --  DB null 2
08.04.2023 08:41:34:758 -- SelectCurrentDbActivity.OnPause 2
08.04.2023 08:41:34:764 -- FileSelect.OnCreate
08.04.2023 08:41:34:793 -- FileSelect.OnStart
08.04.2023 08:41:34:887 -- PasswordActivity.OnCreate 3
08.04.2023 08:41:34:888 -- PasswordActivity:apptask= 3
08.04.2023 08:41:34:989 -- GetIocFromLaunchIntent()
08.04.2023 08:41:34:990 -- no keyprovider specified
08.04.2023 08:41:34:991 -- Reset keyfile

**Comment by thwerks** [comment]
I'm really confused. I can't access the settings or debugging options because there's no database loaded. I cant seem to do anything until a database is loaded but i can't load my database. I'm really sure everything is correct. I'm thinking KeePass isn't compatible with my crypto settings or something.

My OpenSSH settings:

Ciphers aes256-gcm@openssh.com
MACs hmac-sha2-512-etm@openssh.com
KexAlgorithms ecdh-sha2-nistp521
Protocol 2

Host key is 16384  bits.  All other android apps supporting SSH i tested work fine. ES File Explorer and JuiceSSH, no issues. 


I set all my crypto settings back to default and made a 2048 bit host key, still nothing. After pressing ok, just drops back to setup screen.  I can't get anything to work no matter what settings i use on my SSH server.  Other apss work no matter what settings i used. 

![Screenshot_20230408-154331](https://user-images.githubusercontent.com/90361336/230710046-fe18d073-4f8e-4d45-aecc-e2bdba42c803.png)

**Comment by Equidamoid** [comment]
Hey @PhilippC 
I've got similar looking problem, and these lines from `auth.log` seem to give a hint:
```
Apr  9 19:38:15 gullinbursti sshd[11105]: userauth_pubkey: signature algorithm ssh-rsa not in PubkeyAcceptedAlgorithms [preauth]
Apr  9 19:38:15 gullinbursti sshd[11105]: error: Received disconnect from xxx port 52286:3: com.jcraft.jsch.JSchException: Auth fail [preauth]
Apr  9 19:38:15 gullinbursti sshd[11105]: Disconnected from authenticating user xxx xxx port 52286 [preauth]
```
Looks like the key pair generated by keepass2android uses old bad algorithm and therefore sshd says no.

Looking at https://github.com/PhilippC/keepass2android/blob/7b31f6ec5aea9cbe540b9e59dc8035523e0f770d/src/java/JavaFileStorage/app/src/main/java/com/jcraft/jsch/jce/SignatureRSAN.java#L46 (and the call chain from `SftpStorage.java`) it seems that keepass2android uses ssh-rsa with sha1 signature (default in jsch), which is indeed old and bad.

Changing RSA to ECDSA/ED.* in https://github.com/PhilippC/keepass2android/blob/d0da83182f0d1ac640c24e6916dd8bc2b6ad2de5/src/java/JavaFileStorage/app/src/main/java/keepass2android/javafilestorage/SftpStorage.java#L383 should fix the problem.

**Comment by hobias** [comment]
I have the same issue. I would very much appreciate an update of the signature algorithm to be more recent and secure!

**Comment by waltersdorff** [comment]
Did you try the latest update? V. 1.09e

From the changelog:

Switch to new SFTP implementation, supporting modern public key algorithms such as rsa-sha2-256

Works for me!

**Comment by hobias** [comment]
Great timing. Thanks for the hint. Updated the app and it works for me too. I think the issue is resolved then.

**Comment by thwerks** [comment]
Confirmed working with version 1.09e-r7 from apkpure.com

With settings:
Ciphers aes256-gcm@openssh.com
MACs hmac-sha2-512-etm@openssh.com
KexAlgorithms ecdh-sha2-nistp521
Protocol 2

My host key is 16384 bits.

Just though i'd mention that while I couldn't get this to work properly with previous version, I did use Syncthing to sync my keepass database from my workstation to my phone which worked well. It might be preferable for some users to do it this way rather than accessing the file via SFTP, because it opens quicker and you have the benefit of having a local cached copy of the file. It still synchronizes as expected though the I dont really know how strong the encryption of Syncthing is. I believe it uses ssh though.
